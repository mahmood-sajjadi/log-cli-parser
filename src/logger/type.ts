export interface logger {
  log: (str: string) => void;
  error: (str: string) => void;
}
