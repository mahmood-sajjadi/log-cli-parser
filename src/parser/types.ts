type IP = string;
type User = string;
type Method = string;

export interface AccessLog {
  ip: IP;
  remoteLogName: string;
  user: User;
  accessTime: Date;
  Request: {
    method: Method;
    uri: string;
    protocol: string;
  }
  statusCode: number;
  responseSize: number | '-';
  agent: string;
}

export type LineParser = (data: string) => AccessLog;

/**
 * @deprecated Use `LineParser` instead. This function will be removed in version 2.0.
 */
export type Parser = (data: string) => AccessLog[];
