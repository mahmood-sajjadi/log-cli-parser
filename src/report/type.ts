export type ReportFromStream<L, I> = (list: L, item: I) => L;
