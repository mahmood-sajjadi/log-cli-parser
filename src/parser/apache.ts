import type { Parser, AccessLog } from './types';

export const ApacheLogLineParser = (line: string): AccessLog => {
  const logPattern = /^(\S+) (\S+) (\S+) \[([^\]]+)\] "(\S+)\s(\S+)\s(\S+)" (\d{3}) (\d+|-) "[^"]+" "([^"]+)"/;

  const match = line.match(logPattern);
  if (!match) {
    throw new Error("Invalid log format");
  }

  const [
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _,
    ip,
    logName,
    user,
    datetime,
    method,
    uri,
    protocol,
    status,
    size,
    agent,
  ] = match;

  return {
    accessTime: new Date(datetime),
    agent,
    ip,
    remoteLogName: logName,
    Request: {
      method,
      protocol,
      uri,
    },
    responseSize: size === '-' ? '-' : Number.parseInt(size, 10),
    statusCode: Number.parseInt(status, 10),
    user,
  }
}

/**
 * @deprecated Use `ApacheLogLineParser` instead. This function will be removed in version 2.0.
 */
export const ApacheLogFileParserFactory = (delimiter = '\n') => {
  const ApacheLogFileParser: Parser = (data: string) => {
    return data.split(delimiter).map(line => ApacheLogLineParser(line));
  }
  return ApacheLogFileParser;
}
