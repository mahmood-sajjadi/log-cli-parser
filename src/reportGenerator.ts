import { readLogFileByLine } from './handler/file';
import { ApacheLogLineParser } from './parser/apache';
import { countInStream } from './report/count';
import type { countResult } from './report/count';
import { distinctInStream } from './report/distinct';
import type { Logger } from './logger/type';
import type { AccessLog, LineParser } from './parser/types';

export interface RawReport {
  distinctIps: string[];
  visitedUrlsCount: countResult<string>;
  activityIpsCount: countResult<string>;
}

export function parseLine(rawLineData: string, config: {
  parser: LineParser
}): AccessLog {
  return config.parser(rawLineData);
}

// TODO: to make it more generic, it needs to accept list of reporter functions
// and their configurations. then this function will loop them all to generate
// all the reports
export function addToReport(reportResult: RawReport, data: AccessLog): RawReport {
  // countInStream will mutate the report
  reportResult.visitedUrlsCount = countInStream(reportResult.visitedUrlsCount, data.Request.uri);
  reportResult.activityIpsCount = countInStream(reportResult.activityIpsCount, data.ip);
  // distinctInStream will mutate the report
  reportResult.distinctIps = distinctInStream(reportResult.distinctIps, data.ip);

  return reportResult;
}

export function parseFileAsStream(file: string, customLogger: Logger) {
  return new Promise<RawReport>((resolve) => {
    let report: RawReport = {
      distinctIps: [],
      visitedUrlsCount: [],
      activityIpsCount: [],
    };

    const read = (line: string) => {
      const data = parseLine(line, {
        parser: ApacheLogLineParser,
      });
      // because of inner functions, addToReportFromStream will mutate the report
      report = addToReport(report, data);
    }
    const close = () => {
      resolve(report);
    }
    const error = (err: any) => {
      // TODO: to log the error or handle it better,
      // for now I ignore the error and log it in console only
      customLogger.error(`error on parsing the file: ${err}`);
    }
    readLogFileByLine(file, {
      read,
      close,
      error,
    });
  });
}
