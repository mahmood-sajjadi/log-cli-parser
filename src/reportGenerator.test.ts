import { AccessLog } from './parser/types';
import { addToReport } from './reportGenerator';
import type { RawReport } from './reportGenerator';

describe('addToReport', () => {
  test('WHEN new access log line PROVIDED that should be added to reports', () => {
    const log: AccessLog = {
      "Request": {
        "method": "GET",
        "protocol": "HTTP/1.1",
        "uri": "/intranet-analytics/",
      },
      "accessTime": new Date(),
      "agent": "Mozilla/5.0 (X11; U; Linux x86_64; fr-FR) AppleWebKit/534.7 (KHTML, like Gecko) Epiphany/2.30.6 Safari/534.7",
      "ip": "177.71.128.21",
      "remoteLogName": "-",
      "responseSize": 3574,
      "statusCode": 200,
      "user": "-",
    };
    const result: RawReport = addToReport({
      distinctIps: [],
      visitedUrlsCount: [],
      activityIpsCount: [],
    }, log);
    expect(result).toMatchSnapshot();
  });
});
