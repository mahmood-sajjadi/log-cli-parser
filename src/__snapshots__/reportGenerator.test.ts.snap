// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addToReport WHEN new access log line PROVIDED that should be added to reports 1`] = `
{
  "activityIpsCount": [
    {
      "count": 1,
      "item": "177.71.128.21",
    },
  ],
  "distinctIps": [
    "177.71.128.21",
  ],
  "visitedUrlsCount": [
    {
      "count": 1,
      "item": "/intranet-analytics/",
    },
  ],
}
`;
